### PR TITLE
Included partition_key's length when calculating request size to kinesis. Increased record size limit from 50KB to 1MB

### PR DESCRIPTION
**Pull request by trongnhanbmt**

### CODE REVIEW

**Comment by trongnhanbmt** [comment]
Because we don't count partition_key while calculating request size, there is an edge case when the payload is less than 5MB but kinesis still refuses it because payload + partition_key > 5MB.

Also, the limit for a kinesis record has been increased to 1MB:
http://aws.amazon.com/about-aws/whats-new/2015/06/amazon-kinesis-announces-put-pricing-change-1mb-record-support-and-the-kinesis-producer-library/

File: lib/fluent/plugin/out_kinesis.rb [hunk 1]
```diff
@@ -27,7 +27,7 @@ class OutputFilter < Fluent::BufferedOutput
     USER_AGENT_NAME = 'fluent-plugin-kinesis-output-filter'
     PROC_BASE_STR = 'proc {|record| %s }'
     PUT_RECORDS_MAX_COUNT = 500
-    PUT_RECORD_MAX_DATA_SIZE = 1024 * 50
+    PUT_RECORD_MAX_DATA_SIZE = 1024 * 1024
```
Comment (trongnhanbmt) [inline lib/fluent/plugin/out_kinesis.rb]: Bump the increase from 50KB to 1MB.

File: lib/fluent/plugin/out_kinesis.rb [hunk 2]
```diff
@@ -115,7 +115,7 @@ def write(chunk)
         unless record_exceeds_max_size?(record['data'])
           true
         else
-          log.error sprintf('Record exceeds the 50KB per-record size limit and will not be delivered: %s', record['data'])
+          log.error sprintf('Record exceeds the %.3f KB(s) per-record size limit and will not be delivered: %s', PUT_RECORD_MAX_DATA_SIZE / 1024.0, record['data'])
```
Comment (trongnhanbmt) [inline lib/fluent/plugin/out_kinesis.rb]: Make the error message more dynamic.

File: lib/fluent/plugin/out_kinesis.rb [hunk 3]
```diff
@@ -222,15 +222,16 @@ def build_records_array_to_put(data_list)
       records_payload_length = 0
       data_list.each{|data_to_put|
         payload = data_to_put[:data]
-        if records.length >= PUT_RECORDS_MAX_COUNT or (records_payload_length + payload.length) >= PUT_RECORDS_MAX_DATA_SIZE
+        partition_key = data_to_put[:partition_key]
+        if records.length >= PUT_RECORDS_MAX_COUNT or (records_payload_length + payload.length + partition_key.length) >= PUT_RECORDS_MAX_DATA_SIZE
```
Comment (trongnhanbmt) [inline lib/fluent/plugin/out_kinesis.rb]: included partition_key in calculating request size.

File: lib/fluent/plugin/out_kinesis.rb [hunk 4]
```diff
@@ -222,15 +222,16 @@ def build_records_array_to_put(data_list)
...
           records_array.push(records)
           records = []
           records_payload_length = 0
         end
         records.push(data_to_put)
-        records_payload_length += payload.length
+        records_payload_length += (payload.length + partition_key.length)
       }
-      records_array.push(records)
+      records_array.push(records) unless records.empty?
```
Comment (trongnhanbmt) [inline lib/fluent/plugin/out_kinesis.rb]: Don't push empty array to an array.

File: test/plugin/test_out_kinesis.rb [hunk 5]
```diff
@@ -312,75 +312,99 @@ def test_get_key
 
   def test_record_exceeds_max_size
     d = create_driver
+
+    # PUT_RECORD_MAX_DATA_SIZE = 1024*1024 is way too big, try something smaller (10), just to verify the logic not the actual value
+    original_put_record_max_data_size = d.instance.class.send(:remove_const, :PUT_RECORD_MAX_DATA_SIZE) if d.instance.class.const_defined?(:PUT_RECORD_MAX_DATA_SIZE)
```
Comment (trongnhanbmt) [inline test/plugin/test_out_kinesis.rb]: Separate the logic to verify the limits with actual limit values.

**Comment by imaifactory** [comment]
Thanks! I will look into this.

**Comment by trongnhanbmt** [comment]
@imaifactory Did you have a chance to look at this? I deployed this fix to my local env and it worked well so far.

**Comment by imaifactory** [comment]
@trongnhanbmt 

Thank you for your contribution! And sorry for long lag. This is really great work.
Please let me confirm one thing: I think PutRecords record size only counts `Data` attribute.
http://docs.aws.amazon.com/ja_jp/kinesis/latest/APIReference/API_Record.html
I'm talking with Kinesis team to clarify this.

After that, I will let you know. If it counts all attributes, I will proceed this, and if not, I will ask you to fix around that.

Thanks again!

**Comment by imaifactory** [comment]
Ah, sorry it was my mis-reading. The document clearly says:
> When the data blob (the payload before base64-encoding) is added to the partition key size, the total size must not exceed the maximum record size (1 MB).

I will proceed the pipeline, thanks.

**Comment by imaifactory** [comment]
@trongnhanbmt 
Thanks for waiting. We can accept this pull request now.
Please confirm that this change is provided under Apache License 2.0?

**Comment by trongnhanbmt** [comment]
I confirmed. Thanks much @imaifactory

**Comment by imaifactory** [comment]
Thank you!

**Pull request merged**